linkerd/protocol/h2/src/main/scala/io/buoyant/linkerd/protocol/h2/IstioIngressIdentifier.scala

```diff
@@ -0,0 +1,128 @@
+import io.buoyant.k8s.istio._
+import io.buoyant.k8s.{ClientConfig, IngressCache}
+import io.buoyant.linkerd.IdentifierInitializer
+import io.buoyant.linkerd.protocol.H2IdentifierConfig
+import io.buoyant.linkerd.protocol.h2.ErrorReseter.H2ResponseException
+import io.buoyant.router.RoutingFactory.{BaseDtab, DstPrefix, IdentifiedRequest, Identifier, RequestIdentification, UnidentifiedRequest}
+import istio.proxy.v1.config.HTTPRedirect
+
+class IstioIngressIdentifier(
+  val pfx: Path,
+  baseDtab: () => Dtab,
+  namespace: Option[String],
+  apiClient: Service[http.Request, http.Response],
+  annotationClass: String,
+  val routeCache: RouteCache,
+  val clusterCache: ClusterCache
+) extends Identifier[Request] with IstioIdentifierBase[Request] {
+
+  private[this] val ingressCache = new IngressCache(namespace, apiClient, annotationClass)
+
+  override def apply(req: Request): Future[RequestIdentification[Request]] = {
+    val matchingPath = ingressCache.matchPath(Some(req.authority), req.path)
+    matchingPath.flatMap {
+      case None =>
+        Future.value(new UnidentifiedRequest(s"no ingress rule matches ${req.authority}:${req.path}"))
+      case Some(ingressPath) =>
+        val clusterName = s"${ingressPath.svc}.${ingressPath.namespace}.svc.cluster.local"
+
+        // use clusterCache to transform any port Int into a port name
+        val portName = Try(ingressPath.port.toInt).toOption match {
+          case Some(portNumber) => clusterCache.get(s"$clusterName:$portNumber").map {
+            case Some(Cluster(_, p)) => Some(p)
+            case None => None
+          }
+          case None => Future.value(Some(ingressPath.port))
+        }
+
+        Future.join(portName, routeCache.getRules).flatMap {
+          case (Some(port), rules) =>
+            val filteredRules = filterRules(rules, clusterName, reqToMeta(req))
+            (maxPrecedenceRule(filteredRules) match {
+              case Some((ruleName, rule)) =>
+                rule.`redirect` match {
+                  case Some(redir) => redirectRequest(redir, req)
+                  case None =>
+                    val (uri, authority) = httpRewrite(rule, reqToMeta(req))
+                    rewriteRequest(uri, authority, req)
+                    Future.value(pfx ++ Path.Utf8("route", ruleName, port))
+                }
+              //forward requests which have no matching rules to an empty label selector
+              case None => Future.value(pfx ++ Path.Utf8("dest", clusterName, "::", port))
+            }).map { path =>
+              val dst = Dst.Path(path, baseDtab(), Dtab.local)
+              new IdentifiedRequest(dst, req)
+            }
+          case _ => Future.value(new UnidentifiedRequest(s"ingress path ${ingressPath.svc}:${ingressPath.port} does not match any istio vhosts"))
+        }
+    }
+  }
+
+  def redirectRequest(redir: HTTPRedirect, req: Request): Future[Nothing] = {
+    val resp = Response(Status.Found, Stream.empty(0))
```

Comment:
Bleh, yeah thanks for looking at this, it's a different Stream. Will update to h2.Stream